build time increased after updating from plugin version 4.1.2 to 4.2.x

### Expected Behavior

the build takes the same (or less) amount of time after plugin version upgrade

### Actual Behaviour

we have a pretty big multi-module project where build time increased significantly after updating the Gradle plugins to version 4.2.1 from 4.1.2. Our CI/CD build time went up from ~20 mins to ~50. Local builds also got slower, although not that drastically, but for example the simplest task also takes at least 30 secs (was < 1s before), because the configuration phase takes several seconds.

### Steps To Reproduce

_No response_

### Environment Information

Local:
- MacOS 12.7.1
- Gradle 8.4
- JVM 17.0.5 (Eclipse Adoptium 17.0.5+8)

CI/CD:
- ubuntu-2004:2023.07.1
- Gradle 8.4
- JVM openjdk-17-jre

### Example Application

_No response_

### Version

Plugin v4.2.1, Micronaut v4.2.4

@melix we're seeing errors in non-multiplatform projects too on 4.3.0. the interesting part of the stacktrace could be this one probably:

```
Caused by: org.gradle.api.internal.plugins.PluginApplicationException: Failed to apply plugin class 'io.micronaut.gradle.MicronautComponentPlugin'.
	at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:176)
	at org.gradle.api.internal.plugins.DefaultPluginManager.apply(DefaultPluginManager.java:153)
	at io.micronaut.gradle.MicronautMinimalLibraryPlugin.apply(MicronautMinimalLibraryPlugin.java:32)
	at io.micronaut.gradle.MicronautMinimalLibraryPlugin.apply(MicronautMinimalLibraryPlugin.java:27)
	at org.gradle.api.internal.plugins.ImperativeOnlyPluginTarget.applyImperative(ImperativeOnlyPluginTarget.java:43)
	at org.gradle.api.internal.plugins.RuleBasedPluginTarget.applyImperative(RuleBasedPluginTarget.java:51)
	at org.gradle.api.internal.plugins.DefaultPluginManager.addPlugin(DefaultPluginManager.java:190)
	at org.gradle.api.internal.plugins.DefaultPluginManager.access$100(DefaultPluginManager.java:54)
```

--- a/minimal-plugin/src/main/java/io/micronaut/gradle/internal/AutomaticDependency.java
+++ b/minimal-plugin/src/main/java/io/micronaut/gradle/internal/AutomaticDependency.java
@@ -42,8 +42,10 @@ public record AutomaticDependency(
         String coordinates,
         Optional<ConfigurableVersionProperty> versionProperty
 ) {
+
     public void applyTo(Project p) {
         p.getPlugins().withType(MicronautComponentPlugin.class, unused -> {
+            p.afterEvaluate(unusedProject -> VersionCatalogLookupCache.get().clear());
             var dependencyHandler = p.getDependencies();
             var micronautExtension = p.getExtensions().getByType(MicronautExtension.class);
             var ignoredDependencies = micronautExtension.getIgnoredAutomaticDependencies();
@@ -65,13 +67,11 @@ public record AutomaticDependency(
                     if (versionCatalogs != null) {
                         Optional<VersionCatalog> mn = versionCatalogs.find("mn");
                         if (mn.isPresent()) {
+                            // The cache should ideally use Gradle build services, but that's currently impossible due to
+                            // https://github.com/gradle/gradle/issues/17559
+                            var catalogCache = VersionCatalogLookupCache.get();
                             VersionCatalog micronautCatalog = mn.get();
-                            Optional<Provider<MinimalExternalModuleDependency>> dependencyProvider = micronautCatalog.getLibraryAliases()
-                                .stream()
-                                .map(micronautCatalog::findLibrary)
-                                .map(Optional::get)
-                                .filter(d -> coordinates.equals(d.get().getModule().toString()))
-                                .findFirst();
+                            Optional<Provider<MinimalExternalModuleDependency>> dependencyProvider = catalogCache.findDependencyFromCatalog(micronautCatalog, coordinates);
                             if (dependencyProvider.isPresent()) {
                                 return List.of(dependencyProvider.get().get());
                             }

--- a/minimal-plugin/src/main/java/io/micronaut/gradle/internal/VersionCatalogLookupCache.java
+++ b/minimal-plugin/src/main/java/io/micronaut/gradle/internal/VersionCatalogLookupCache.java
@@ -1,0 +1,77 @@
+/*
+ * Copyright 2003-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.gradle.internal;
+
+import org.gradle.api.artifacts.MinimalExternalModuleDependency;
+import org.gradle.api.artifacts.VersionCatalog;
+import org.gradle.api.provider.Provider;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This class is used as a workaround for very poor performance
+ * of version catalog scanning in Gradle.
+ * See https://github.com/micronaut-projects/micronaut-gradle-plugin/issues/932
+ * and https://github.com/gradle/gradle/pull/27444/files
+ */
+class VersionCatalogLookupCache implements AutoCloseable {
+    private final AtomicReference<VersionCatalog> oldCatalog = new AtomicReference<>();
+    private final Map<String, Provider<MinimalExternalModuleDependency>> coordsToDependencyProvider = new ConcurrentHashMap<>();
+
+    static VersionCatalogLookupCache get() {
+        class Holder {
+            private static final VersionCatalogLookupCache INSTANCE = new VersionCatalogLookupCache();
+        }
+        return Holder.INSTANCE;
+    }
+
+    private VersionCatalogLookupCache() {
+
+    }
+
+    void clear() {
+        oldCatalog.set(null);
+        coordsToDependencyProvider.clear();
+    }
+
+    private void memoize(VersionCatalog catalog) {
+        if (oldCatalog.compareAndSet(null, catalog)) {
+            coordsToDependencyProvider.clear();
+            catalog.getLibraryAliases()
+                .stream()
+                .parallel()
+                .forEach(alias -> {
+                    var library = catalog.findLibrary(alias);
+                    library.ifPresent(lib ->
+                        coordsToDependencyProvider.put(lib.get().getModule().toString(), lib)
+                    );
+                });
+        }
+    }
+
+    public Optional<Provider<MinimalExternalModuleDependency>> findDependencyFromCatalog(VersionCatalog catalog, String coordinates) {
+        memoize(catalog);
+        return Optional.ofNullable(coordsToDependencyProvider.get(coordinates));
+    }
+
+    @Override
+    public void close() {
+        coordsToDependencyProvider.clear();
+    }
+}

